language/move-prover/bytecode/src/access_path.rs

```diff
@@ -636,36 +646,32 @@ impl AccessPath {
     /// Bind free type variables to `type_actuals`.
     pub fn substitute_footprint(
         &self,
-        actuals: &[AbsAddr],
         type_actuals: &[Type],
+        temp_indices: &[TempIndex],
+        func_env: &FunctionEnv,
         sub_map: &dyn AccessPathMap<AbsAddr>,
     ) -> AbsAddr {
-        let mut acc = AbsAddr::default();
         let mut new_offsets = self.offsets.clone();
         new_offsets.iter_mut().for_each(|o| {
             o.substitute_footprint(type_actuals);
         });
         match &self.root {
             Root::Formal(i) => {
-                let actual = &actuals[*i];
-                if !actual.is_empty() {
-                    let new_ap = AccessPath::new(self.root.clone(), new_offsets);
-                    acc.join(&new_ap.prepend_addrs(actual));
-                } else {
-                    // TODO: change actuals to &[TempIndex], look up actuals[i]/new_offsets in sub_map")
-                }
+                let temp_index = temp_indices[*i];
+                let root = Root::from_index(temp_index, func_env);
+                let ap = AccessPath::new(root, new_offsets);
+                sub_map.get_access_path(ap).cloned().unwrap_or_default()
```

Comment:
This doesn't look quite right to me. At a high level, what this code is doing is:

- `self` is a footprint path in the callee state that reads from the environment (e.g., `Formal(0)/f/g)`.
- We want to resolve the formal in the caller environment (`sub_map`), then prepend the result to the callee's path. For example, if `Formal(0)` is bound to `Formal(1)/e` in the caller, we would like this function to produce the path `Formal(1)/e/f/g`. In the old code, this is accomplished via `prepend_addrs(Formal(0)/f/g, Formal(1)/e)`. But the new code drops the call to `prepend_addrs`.

Reply:
This was slightly nastier to fix than I hoped but I believe the latest version should cover the cases that did not work in both the previous version of this PR and the previous implementation.
At least, it does seem to fix the issue mentioned here: https://github.com/diem/diem/pull/8706/files#r666360471